walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,83 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Sort list by events.
+    ArrayList<Event> eventList = new ArrayList(events);
+    Collections.sort(eventList, (Event e1, Event e2) -> TimeRange.ORDER_BY_START.compare(e1.getWhen(), e2.getWhen()));
+    
+    Collection<TimeRange> optionalIncludedOpenTimes = getOpenTimes(eventList, request, /* optionalAttendees= */ true);
+    Collection<TimeRange> requiredOpenTimes = getOpenTimes(eventList, request, /* optionalAttendees= */ false);
+
+    if (optionalIncludedOpenTimes.size() > 0) {
+      return optionalIncludedOpenTimes;
+    } 
+    else if (request.getAttendees().size() > 0) {
+      return requiredOpenTimes;
+    } 
+    else {
+      return Collections.emptyList();
+    }
   }
+
+  /**
+   * A helper method that will get all the open times in a day for a meeting to occur.
+   * @param events All the events happening in the day.
+   * @param request The meeting request that will be scheduled sometime in the day.
+   * @param optionalAttendees A choice on whether to schedule around optional attendees as well.
+   * @return A list of all the possible times the meeting request can be scheduled.
+   */
+  private Collection<TimeRange> getOpenTimes(Collection<Event> events, MeetingRequest request, boolean optionalAttendees){
+    ArrayList<TimeRange> openTimes = new ArrayList<TimeRange>();
+
+    // Edge case when a meeting is longer than a full day.
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()){
+      return openTimes;
+    }
+    
+    int dayStart = TimeRange.START_OF_DAY;
+
+    for (Event event: events) {
+      // Ignore the event if no required members are in it.
+      boolean noRequiredMembers = Collections.disjoint(event.getAttendees(), request.getAttendees());
+      boolean noOptionalMembers = Collections.disjoint(event.getAttendees(), request.getOptionalAttendees());
+
+      if (optionalAttendees){
+        if (noRequiredMembers && noOptionalMembers){
+          continue;
+        }     
+      }
+      else if (noRequiredMembers) {
+        continue;
+      }
+
```

Comment:
nit: instead of having ```@param request``` you could pass ```Collection<String> all attendees``` which is optional + mandatory attendees, and pass ```meeting duration```. This can help reduce your condition on **lines 48 to 71** to a single if since you would be considering all attendees at the same time